### PR TITLE
add ego1 fpga board support and multiple i2c / spi / uart port in qmtech wukong board

### PR DESCRIPTION
**Pull request by SanadaShinken**
Dear Sir :

two board support are added.

1. add one new board support, ego1. there is no ddr or sdram on board, olnly sram. the main memory is based on fpga internal block memory.

2. modified the qmtech wukong board for multiple i2c / spi / uart.  

BR, Sanada

### CODE REVIEW

File: litex/soc/software/bios/cmds/cmd_bios.c [hunk 1]
```diff
@@ -8,10 +8,61 @@
...
+	(0x01 | 0x02 | 0x04 | 0x08 | 0x10 | 0x20 | 0x40), // 8
+	(0x01 | 0x02 | 0x04 | 0x08 | 0x20 | 0x40),        // 9
+	(0x01 | 0x02 | 0x04 | 0x10 | 0x20 | 0x40),        // A
+	(0x04 | 0x08 | 0x10 | 0x20 | 0x40),               // B
+	(0x01 | 0x08 | 0x10 | 0x20),                      // C
+	(0x02 | 0x04 | 0x08 | 0x10 | 0x40),               // D
+	(0x01 | 0x08 | 0x10 | 0x20 | 0x40),               // E
+	(0x01 | 0x10 | 0x20 | 0x40),                      // F
+};
+
```
Comment (enjoy-digital) [inline litex/soc/software/bios/cmds/cmd_bios.c:47]: This code seems related to the seven segment, so should probably not be in cmd_bios.

File: litex/soc/software/bios/cmds/cmd_bios.c [hunk 2]
```diff
@@ -8,10 +8,61 @@
...
+	unsigned int i = (1000 * counts * ticks_per_delay) / ticks_per_us;
+	while(i--) {__asm__ volatile("nop");}
+}
+
+static void delay_us(unsigned int counts) {	// micro second based
+	unsigned int i = (counts * ticks_per_delay) / ticks_per_us;
+	while(i--) {__asm__ volatile("nop");}
+}
+
 /**
```
Comment (enjoy-digital) [inline litex/soc/software/bios/cmds/cmd_bios.c:66]: We already have `busy_wait` and `cdelay` functions in the BIOS, we should try to avoid introducing others unless if really required.

File: litex/soc/software/bios/cmds/cmd_bios.c [hunk 3]
```diff
@@ -56,6 +107,56 @@ static void ident_handler(int nb_params, char **params)
...
+	printf("start: %u\r\n", start);
+	printf("end:   %u\r\n", end);
+	printf("elapsed ticks: %u\r\n", tick_counts);
+	printf("ticks per delay: %u\r\n", ticks_per_delay);
+	printf("ticks per microsecond: %u\r\n", ticks_per_us);
+}
+
+define_command(delay_time, delay_time_handler, "delay time of the system", SYSTEM_CMDS);
+
+
```
Comment (enjoy-digital) [inline litex/soc/software/bios/cmds/cmd_bios.c:158]: This could be  a nice example of use of the timer, but I'm still not sure it's a useful feature in the BIOS, I'll think about it.

File: litex/soc/software/bios/cmds/cmd_bios.c [hunk 4]
```diff
@@ -161,6 +262,120 @@ static void leds_handler(int nb_params, char **params)
...
+	}
+	seg_val = seg_val & 0x0f;
+	seg_val = (seg_pos < 4) ? (hex_val[seg_val]) : ((hex_val[seg_val]) << 8);
+	seven_seg_out_write(seg_val);
+
+	printf("Settings Seg@%d to 0x%x", seg_pos, seg_val);
+}
+
+define_command(seven_seg, seven_seg_handler, "Set Seven Seg Leds value", SYSTEM_CMDS);
+
```
Comment (enjoy-digital) [inline litex/soc/software/bios/cmds/cmd_bios.c:298]: Adding Seven Segment support should probably be a separate PR.

File: litex/soc/software/bios/cmds/cmd_bios.c [hunk 5]
```diff
@@ -161,6 +262,120 @@ static void leds_handler(int nb_params, char **params)
...
+				}
+			}
+		}
+	}
+}
+
+define_command(seven_seg_counter, seven_seg_counter_handler, "Seven Seg Leds Counter", SYSTEM_CMDS);
+#endif
+
 /**
```
Comment (enjoy-digital) [inline litex/soc/software/bios/cmds/cmd_bios.c:379]: Similar to the delay function, I would rather integrated this in the Demo app.

File: litex/soc/software/bios/cmds/cmd_bios.c [hunk 6]
```diff
@@ -203,3 +418,50 @@ static void cmd_sim_mark_handler(int nb_params, char **params)
...
+		for (int x = 0; x < 80; x++) {
+			vga[0] = (input_val & 0xFF);
+			vga[1] = color;
+			vga += 2;
+		}
+	}
+}
+define_command(vga_test, vga_test_handler, "VGA Test", SYSTEM_CMDS);
+#endif
+
```
Comment (enjoy-digital) [inline litex/soc/software/bios/cmds/cmd_bios.c:467]: This would also be a nice example for the Demo App.

File: litex/tools/litex_json2dts.py [hunk 7]
```diff
@@ -192,6 +195,34 @@ def generate_dts(d, initrd_start=None, initrd_size=None, polling=False):
...
+                device_type = "serial";
+                compatible = "litex,liteuart";
+                reg = <0x{uart_csr_base:x} 0x100>;
+                {uart_interrupt}
+                status = "okay";
+            }};
+""".format(
+    uart_csr_base  = d["csr_bases"]["uart2"],
+    uart_interrupt = "" if polling else "interrupts = <{}>;".format(d["constants"]["uart2_interrupt"]))
+
```
Comment (enjoy-digital) [inline litex/tools/litex_json2dts.py:225]: We are discussing this in https://github.com/enjoy-digital/litex/pull/777, for now it's still not decided if multiple instance of the same peripheral will be directly support in LiteX and preliminary work as described in #777 is required for that.

File: litex/tools/litex_json2dts.py [hunk 8]
```diff
@@ -338,11 +369,35 @@ def generate_dts(d, initrd_start=None, initrd_size=None, polling=False):
...
+                litespi,num-cs = <1>;
+
+                #address-cells = <1>;
+                #size-cells    = <0>;
+
+                spidev1: spidev@0 {{
+                    compatible = "linux,spidev";
+                    reg = <0>;
+                    spi-max-frequency = <2000000>;
                     status = "okay";
```
Comment (enjoy-digital) [inline litex/tools/litex_json2dts.py:397]: Same comment than for UART.

File: litex/tools/litex_json2dts.py [hunk 9]
```diff
@@ -355,6 +410,24 @@ def generate_dts(d, initrd_start=None, initrd_size=None, polling=False):
...
+    if "i2c2" in d["csr_bases"]:
+        dts += """
+            i2c2: i2c@{i2c2_csr_base:x} {{
+                compatible = "litex,i2c";
+                reg = <0x{i2c2_csr_base:x} 0x5>;
+                status = "okay";
+            }};
+""".format(i2c2_csr_base=d["csr_bases"]["i2c2"])
+
     # XADC -----------------------------------------------------------------------------------------
```
Comment (enjoy-digital) [inline litex/tools/litex_json2dts.py:431]: Same comment than for UART.

**Review by enjoy-digital — CHANGES_REQUESTED** [review]
Thanks a lot @SanadaShinken for the PR. There are several interesting additions/features that we can merge but I'll need to have separate PR since this PR is mixing too much additions/features, can you make a separate PR for each of these points:
- Base support for Seven Segment display would be a nice addition to the BIOS, but the BIOS would only support simple control (your seven_seg <seg_pos> <seg_val) command).
- More fancy support command for the Seven Segment (count) could be integrated in the demo app: https://github.com/enjoy-digital/litex/tree/master/litex/soc/software/demo
- The VGA terminal command would also be a nice addition the demo app.
- The Delay command would also be a nice addition to the demo app, providing a concrete example to configure and use the timer.

Your changes to `litex_json2dts.py` seem very specific to your project, it will not be possible to integrate them now (or just the one that would really be generic), so I'll encourage you to use an out-of-tree version of` litex_json2dts.py` in your project and we'll also try to support the different features progressively in upstream LiteX.

File: litex/soc/software/bios/cmds/cmd_bios.c (hunk 1 above)
Comment (SanadaShinken) [reply litex/soc/software/bios/cmds/cmd_bios.c:47]: Dear Sir:

I think what you talk about is right. could I withdraw this PR ? 
after put the code into where you suggest and send PR again?

BR, Sanada

**Comment by SanadaShinken** [comment]
Dear Sir:

I'll close this pull request. and send a new one, only for qmtech wukong.

BR, Sanada
